Enable multiple disk scheduling

JacieChao opened the pull request:
For now, just add Node CRD for multiple disk configuration and add a replica scheduler to schedule replica to specified disk on node.

yasker commented:
@JacieChao 

Please separate your commit into following commits:
1. Add CRD to https://github.com/rancher/longhorn-manager/blob/master/k8s/pkg/apis/longhorn/v1alpha1/types.go
2. Generate the CRD using the code generator
3. Add related fields to datastore
4. Add related fields to API.
5. Other logics.

Make sure after each commit, the manager can be built.

You can split 5 into more commits if reasonable, or switch position with 4.

yasker commented:
Do you need the scheduler to be a controller? I think a function call from replica_controller should be enough. What do you think?

JacieChao commented:
@yasker  Sure, I think it's enough for now. At first I want to create a controller to deal with replica scheduling alone. But I think it's easier to be called from replica_controller. I will remove the unused code in scheduler and separate my commits.

yasker commented on api/model.go:
```diff
@@ -120,6 +120,26 @@ type EngineUpgradeInput struct {
+	Disks           map[string]types.Disk `json:"disks"`
+}
+
+type DiskInput struct {
```
I think it's maybe unnecessary to separate API for add/delete disk.

Considering in the UI, normally it is not going to call API every time when a disk was added or removed. It's easy for UI to call backend that the things have been determined. Also it's easier for the backend to store it in one transaction. And it's a rarely modified parameter, performance shouldn't be a problem. So it should be OK to just use `UpdateNode` to perform the update on all the fields of the node.

So many code below can be simplied.

yasker commented on datastore/kubernetes.go:
```diff
@@ -147,3 +147,22 @@ func (s *DataStore) DeleteEngineImageDaemonSet(name string) error {
 	return nil
 }
+
+func (s *DataStore) GetManagerNode() ([]string, error) {
```
Just use `GetManagerNodeIPMap` with the key as the node name.

yasker commented on datastore/longhorn.go:
```diff
@@ -437,3 +439,61 @@ func (s *DataStore) ListEngineImages() (map[string]*longhorn.EngineImage, error)
+			return nil, err
+		}
+		if node == nil {
+			node, err = s.CreateDefaultNode(nodeName)
```
It's weird to have `GetNodeList` to have a side-effect to create the nonexisting nodes. You can have a manager to create its own node if it doesn't exist when starting up at e.g. https://github.com/rancher/longhorn-manager/blob/master/app/daemon.go#L100 .

yasker commented on manager/volume.go:
```diff
@@ -345,3 +345,27 @@ func (m *VolumeManager) EngineUpgrade(volumeName, image string) error {
+	}
+	if node == nil {
+		// create node for default path
+		return m.ds.CreateDefaultNode(name)
```
Same, don't write as a side-effect of reading.

yasker commented on datastore/longhorn.go:
```diff
@@ -437,3 +439,61 @@ func (s *DataStore) ListEngineImages() (map[string]*longhorn.EngineImage, error)
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: types.NodeSpec{
```
It's better add a finalizer, so we can handle the deletion (which we won't allow for the user but node does go down (and take manager with it) sometime and we need to react to that).

yasker commented on api/router.go:
```diff
@@ -97,6 +97,17 @@ func NewRouter(s *Server) *mux.Router {
 	r.Methods("GET").Path("/v1/hosts").Handler(f(schemas, s.NodeList))
 	r.Methods("GET").Path("/v1/hosts/{id}").Handler(f(schemas, s.NodeGet))
 
+	r.Methods("GET").Path("/v1/nodes").Handler(f(schemas, s.MountNodeList))
```
The old `/v1/hosts` will be removed later, though we need some UI change to accommodate this. You can use the name `NodeList` and `NodeGet`, just rename the previous one to `HostList`/`HostGet` (in a separate commit). MountNode sounds weird.

yasker commented on controller/replica_controller.go:
```diff
@@ -270,6 +273,11 @@ func (rc *ReplicaController) syncReplica(key string) (err error) {
 		}
 	}
+	// check whether the replica need to be scheduled
+	err = rc.scheduler.ScheduleReplica(replica)
```
I think there is a problem with this approach.

Not sure if you observed it, but I think it may result in sometimes two replicas are scheduled on the same host while the other host has none. Since there are multiple workers working to update the replicas simultaneously, you may end up with multiple replicas was scheduled at the exact same time, and choosing the same node for the multiple replicas because when the schedule started, they cannot see the end result of others.

It's more reliable to do scheduling in the volume controller when we create the replica for the first time. It's in sequence for a single volume. https://github.com/rancher/longhorn-manager/blob/master/controller/volume_controller.go#L629

yasker commented on scheduler/replica_scheduler.go:
```diff
@@ -0,0 +1,94 @@
+		if err != nil {
+			return nil
+		}
+		rchecker := checker.NewReplicaChecker(nodeIPMap, replicas)
```
It looks unnecessary to create another package `checker`. It's only one function call we needed.

Make it simple for now. We can expand if it's necessary.

yasker commented on scheduler/replica_scheduler.go:
```diff
@@ -0,0 +1,94 @@
+
+		// TODO Need to add capacity.
+		// Just make sure replica of the same volume be scheduled to different nodes for now.
+		nodeMap := rchecker.ReplicasAffinity()
```
Use a better name, e.g. `preferredNodes`. It looks too similar to `nodeIPMap` which can cause confusion.

yasker commented:
Now I think we need a node controller and node status as well. See updated design.

JacieChao commented:
@yasker NodeController is still working in progress. I will add a new commit when I finish it.

yasker commented on api/model.go:
```diff
@@ -120,6 +120,19 @@ type EngineUpgradeInput struct {
+	client.Resource
+	Name            string                    `json:"name"`
+	AllowScheduling bool                      `json:"allowScheduling"`
+	Disks           map[string]types.DiskSpec `json:"disks"`
```
Define a new type Disk in the API to cover both DiskSpec and DiskStatus.

JacieChao commented:
@yasker I have added unit test for replica_scheduler and updated unit test case for volume_controller and replica_controller. I will add test cases in longhorn-tests later.

yasker commented on controller/replica_controller.go:
```diff
@@ -371,24 +371,10 @@ func (rc *ReplicaController) CreatePodSpec(obj interface{}) (*v1.Pod, error) {
 	// will pin it down to the same host because we have data on it
 	if r.Spec.NodeID != "" {
```
r.Spec.NodeID cannot be "" after your scheduling change.

yasker commented on controller/replica_controller.go:
```diff
@@ -371,24 +371,10 @@ func (rc *ReplicaController) CreatePodSpec(obj interface{}) (*v1.Pod, error) {
+	}
+	// error out if NodeID and DataPath wasn't filled in scheduler
+	if r.Spec.NodeID == "" || r.Spec.DataPath == "" {
+		return nil, fmt.Errorf("There has no avaible node for replica %v", r)
```
Wrong error message. It's a bug.

`fmt.Errorf("BUG: Node or datapath wasn't set for replica %v", r.Name)`

yasker replied:
Pull this functional change into another commit or previous commit. It's kind of been smuggled into a commit suppose to add tests.

yasker commented on scheduler/replica_scheduler_test.go:
```diff
@@ -0,0 +1,231 @@
+	r, err := s.ScheduleReplica(r1)
+	assert.Nil(err)
+	assert.NotNil(r)
+	// assert could not scheduler to node2 and node3
```
Please also add a test for anti-affinity implementation of the scheduler.

yasker commented on scheduler/replica_scheduler.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (rcs *ReplicaScheduler) ScheduleReplica(replica *longhorn.Replica) (*longhorn.Replica, error) {
+	// only called when replica is starting for the first time
```
If this function should only be called when `NodeID == ""`, then error out if NodeID is not empty. It will help to expose the potential bugs.

yasker commented on scheduler/replica_scheduler_test.go:
```diff
@@ -0,0 +1,231 @@
+	}
+}
+
+func TestReplicaScheduler_ScheduleReplica(t *testing.T) {
```
Go style is CamelCase without underscore(`_`). Just do `TestReplicaScheduler` should be fine for now.

yasker commented on controller/volume_controller_test.go:
```diff
@@ -21,6 +21,7 @@ import (
 	lhinformerfactory "github.com/rancher/longhorn-manager/k8s/pkg/client/informers/externalversions"
 
 	. "gopkg.in/check.v1"
+	"k8s.io/api/core/v1"
```
This `k8s.io/api/core/v1` should be put along with other `k8s.io` import above.

yasker commented on scheduler/replica_scheduler_test.go:
```diff
@@ -0,0 +1,290 @@
+package scheduler
+
+import (
+	"fmt"
```
Follow the pattern of other files for import:
```
"fmt"
"testing"
<space>
"k8s.io/xxx"
...
<space>
"github.com/rancher/longhorn-manager/xxxx"
...
<space>
"github.com/rancher/longhorn-manager/k8s/xxx"
...
```

yasker commented on controller/node_controller.go:
```diff
@@ -0,0 +1,228 @@
+		return err
+	}
+	if node == nil {
+		logrus.Debugf("Longhorn node %v does not exist, regenerate a default one", key)
```
I think there are some confusions on how to handle the node deletion. I've updated the design doc for that. Please check `Node creation and deletion` section.

yasker commented on controller/node_controller.go:
```diff
@@ -0,0 +1,228 @@
+	for _, pod := range managerPods {
+		err = nc.syncStatusWithPod(pod, node)
+		if err != nil {
+			return err
```
As I said in the design doc, don't retry if conflict happens, assuming other managers are updating it. I am afraid the requeue may cause a storm of updating conflict in the larger system.

yasker commented on deploy/01-prerequisite/03-crd.yaml:
```diff
@@ -87,3 +87,21 @@ spec:
+    listKind: NodeList
+    plural: nodes
+    shortNames:
+    - lhnode
```
Use `lhn`.

JacieChao commented:
@yasker 
Tested with longhorn-tests. Here's the result
```
+ flake8 .
+ py.test -v .
============================= test session starts ==============================
platform linux2 -- Python 2.7.14, pytest-2.9.2, py-1.5.3, pluggy-0.3.1 -- /usr/local/bin/python
cachedir: .cache
rootdir: /integration/tests, inifile:
collecting ... collected 19 items
test_basic.py::test_hosts_and_settings PASSED
test_basic.py::test_volume_basic PASSED
test_basic.py::test_volume_iscsi_basic PASSED
test_basic.py::test_snapshot PASSED
test_basic.py::test_backup PASSED
test_basic.py::test_volume_multinode PASSED
test_basic.py::test_replica_scheduler PASSED
test_csi.py::test_csi_volume_mount SKIPPED
test_csi.py::test_csi_volume_io SKIPPED
test_driver.py::test_volume_mount PASSED
test_driver.py::test_volume_io PASSED
test_engine_upgrade.py::test_engine_image PASSED
test_engine_upgrade.py::test_engine_offline_upgrade PASSED
test_engine_upgrade.py::test_engine_live_upgrade PASSED
test_engine_upgrade.py::test_engine_image_incompatible PASSED
test_engine_upgrade.py::test_engine_live_upgrade_rollback PASSED
test_ha.py::test_ha_simple_recovery PASSED
test_ha.py::test_ha_salvage PASSED
test_recurring_job.py::test_recurring_job SKIPPED
=================== 16 passed, 3 skipped in 1492.87 seconds ====================
```
I tried to tear down one longhorn-manager to check that Node CRD status set to `down` and not available to schedule replica.

I have updated `README.md` about cleanup node crd before tearing down longhorn system and tested with cleanup process.

yasker commented:
LGTM. Merged, thanks.

yasker commented:
@JacieChao And next time, remember to link your PR to the original issue.
